elfi/methods/diagnostics.py

```diff
@@ -0,0 +1,287 @@
+        # preserve the parameters (thetas) corresponding to the `closest' datasets.
+        thetas = {}
+        E_me = np.inf
+        for set_ss in self.ss_candidates:
+            names_ss = [ss.__name__ for ss in set_ss]
+            thetas_ss = self._obtain_accepted_thetas(set_ss, n_sim, n_acc)
+            thetas[set_ss] = thetas_ss
+            E_ss = self._calc_entropy(thetas_ss, n_acc, k)
+            # If equal, dismiss the combination which contains uninformative summary statistics.
+            if (E_ss == E_me and (len(names_ss_me) > len(names_ss))) or E_ss < E_me:
+                E_me = E_ss
+                names_ss_me = names_ss
+                thetas_closest = thetas_ss[:n_closest]
+            logger.info('Combination %s shows the entropy of %f' % (names_ss, E_ss))
+        logger.info('\nThe minimum entropy of %f was found in %s.\n' % (E_me, names_ss_me))
+
+        # Find the summary-statistics combination with
+        # the minimum mean root sum of squared error (MRRSE).
+        MRSSE_min = np.inf
+        for set_ss in self.ss_candidates:
+            names_ss = [ss.__name__ for ss in set_ss]
+            MRSSE_ss = self._calc_MRSSE(set_ss, thetas_closest, thetas[set_ss])
+            # If equal, dismiss the combination which contains uninformative summary statistics.
+            if (MRSSE_ss == MRSSE_min and (len(names_ss_MRSSE) > len(names_ss))) \
+                    or MRSSE_ss < MRSSE_min:
+                MRSSE_min = MRSSE_ss
+                names_ss_MRSSE = names_ss
+                set_ss_2stage = set_ss
+            logger.info('Combination %s shows the MRSSE of %f' % (names_ss, MRSSE_ss))
+        logger.info('\nThe minimum MRSSE of %f was found in %s.' % (MRSSE_min, names_ss_MRSSE))
+        return set_ss_2stage
+
+    def _obtain_accepted_thetas(self, set_ss, n_sim, n_acc):
+        """Perform the ABC-rejection sampling and identify `closest' parameters.
+
+        The sampling is performed using the initialised simulator.
+
+        Parameters
+        ----------
+        set_ss : array_like
+            Summary-statistics combination to be used in the rejection sampling.
+        n_sim : int
+            Number of the iterations of the rejection sampling.
+        n_acc : int
+            Number of the accepted parameters.
+
+        Returns
+        -------
+        array_like
+            Accepted parameters.
+
+        """
+        # Initialise the distance function.
+        m = self.simulator.model.copy()
+        list_ss = []
+        for ss in set_ss:
+            list_ss.append(elfi.Summary(ss, m[self.simulator.name], model=m))
+        if isinstance(self.fn_distance, str):
+            d = elfi.Distance(self.fn_distance, *list_ss)
+        else:
+            d = elfi.Discrepancy(self.fn_distance, *list_ss)
+
+        # Run the simulations.
+        sampler_rejection = elfi.Rejection(d, batch_size=1, seed=self.seed, pool=self.pool)
```

Comment:
The `batch_size` should be yet another argument.

Reply:
`batch_size` impacts only the computational performance of the diagnostics. Therefore, compared to the user, we have a better idea how to tune it with respect to `n_sim`; also, the additional input argument would increase the function's complexity.

Reply:
A good `batch_size` is very problem specific, so better leave it for the user (can be 1 by default, of course). The performance boost can be huge.